### PR TITLE
Now using OuthToken or username and password to authenticate remote cluster

### LINKED ISSUE #982: Authentication parameters are ignored in OpenShift
##### Issue Overview
Parameters such as `authToken` are ignored or does not work properly when used against remote clusters such as OSIO

##### Expected Behaviour

Log into the remote cluster using auth token provided in the configuration file.

##### Current Behaviour

Token is ignored and if you have not done before an `oc login`, test fails.

### PR DESCRIPTION
**Pull request by dipak-pawar**
#### Short description of what this resolves:
Now using OuthToken or username and password to authenticate remote openshift cluster

#### Changes proposed in this pull request:

- Adds `name`, `password`, `outhToken`, `trustCerts`, `apiVersion` propertis in configuration.
- Adds support for above properties for Kubernetes
- Adds documentation

Fixes #982 , #964

### CODE REVIEW

File: docs/kubernetes.adoc [hunk 1]
```diff
@@ -74,7 +74,12 @@ You can configure Kubernetes by using any of the next configuration properties i
 |===
 | Options                             | Type           | Env | Description                                                   
 | kubernetes.master                   | URL            | Any | The URL to the Kubernetes master                               
-| kubernetes.domain                   | String         | OSE | Domain to use for creating routes for services                 
+| name                                | String         | Any | Username to log in server
```
Comment (lordofthejars) [inline docs/kubernetes.adoc]: Why not username?

File: kubernetes/kubernetes/src/main/java/org/arquillian/cube/kubernetes/impl/ClientCreator.java [hunk 2]
```diff
@@ -34,15 +36,34 @@
     private InstanceProducer<KubernetesClient> producer;
 
     public void createClient(@Observes Configuration config) {
-        if (config.getMasterUrl() != null) {
+
+        if (config.getMasterUrl() == null) {
             producer.set(new DefaultKubernetesClient(new ConfigBuilder()
```
Comment (lordofthejars) [inline kubernetes/kubernetes/src/main/java/org/arquillian/cube/kubernetes/impl/ClientCreator.java]: Create `ConfigBuilder` outside `if` then depending on the values you populate the `ConfigBuilder`, i.e if(master!=null) then set masterUrl; if (token !=null) then set token;...

And then there is no if/else just ifs and at the end you construct the `DefaultKuebernetesClient` and inject in `InstanceProducer`

File: openshift/openshift/src/main/java/org/arquillian/cube/openshift/impl/client/OpenShiftClientCreator.java [hunk 3]
```diff
@@ -24,26 +25,37 @@ public void createClient(@Observes AfterStart afterStart, Configuration conf) {
         CubeOpenShiftConfiguration configuration = (CubeOpenShiftConfiguration) conf;
         System.setProperty("KUBERNETES_TRUST_CERT", "true");
         // override defaults for master and namespace
-        final Config config = new ConfigBuilder()
+        final ConfigBuilder configBuilder = new ConfigBuilder()
```
Comment (lordofthejars) [inline openshift/openshift/src/main/java/org/arquillian/cube/openshift/impl/client/OpenShiftClientCreator.java]: Can you undo the changes of this class? My PR comes with a fix which is a totally different approach by reusing same client instead of instantiating a new one.

File: openshift/openshift/src/main/java/org/arquillian/cube/openshift/impl/fabric8/F8OpenShiftAdapter.java [hunk 4]
```diff
@@ -112,7 +112,7 @@ static OpenShiftConfig toOpenShiftConfig(CubeOpenShiftConfiguration configuratio
             .withMasterUrl(configuration.getKubernetesMaster())
             .withTrustCerts(configuration.isTrustCerts());
 
-        if (configuration.hasOpenshiftBasicAuth()) {
+        if (configuration.hasBasicAuth()) {
```
Comment (lordofthejars) [inline openshift/openshift/src/main/java/org/arquillian/cube/openshift/impl/fabric8/F8OpenShiftAdapter.java]: Same for this class, my PR comes with the final solution by reusing the instance.

File: docs/kubernetes.adoc [hunk 5]
```diff
@@ -74,7 +74,12 @@ You can configure Kubernetes by using any of the next configuration properties i
...
+| password                            | String         | Any | Password to log in server
```
Comment (lordofthejars) [inline docs/kubernetes.adoc]: How about calling parameters `kubernetes.user.name` `kubernetes.user.password` and `kubernetes.user.token`? Any thoughts about this @bartoszmajsak

Comment (bartoszmajsak) [reply docs/kubernetes.adoc]: Do we want to be specific here? Or the idea is to have generic concept of `user` and `pwd` to be applied depends on the target cluster/server/cloud provider (you name it) :)

Comment (bartoszmajsak) [reply docs/kubernetes.adoc]: It might also be confusing, as now OpenShift is k8s, but not everyone knows that?

File: kubernetes/kubernetes/src/main/java/org/arquillian/cube/kubernetes/api/Configuration.java [hunk 6]
```diff
@@ -54,6 +54,12 @@
     String LOGS_COPY = "logs.copy";
     String LOGS_PATH = "logs.path";
 
+    String USERNAME = "name";
```
Comment (bartoszmajsak) [inline kubernetes/kubernetes/src/main/java/org/arquillian/cube/kubernetes/api/Configuration.java]: `username` to match?

File: kubernetes/kubernetes/src/main/java/org/arquillian/cube/kubernetes/impl/ClientCreator.java [hunk 7]
```diff
@@ -34,15 +36,34 @@
...
+                    }
+                });
+
+            if (Strings.isNotNullOrEmpty(config.getToken())) {
+                configBuilder.withOauthToken(config.getToken());
+            }
+
+            if (Strings.isNotNullOrEmpty(config.getUsername()) && Strings.isNotNullOrEmpty(config.getPassword())) {
+                configBuilder.withUsername(config.getUsername());
+                configBuilder.withPassword(config.getPassword());
```
Comment (bartoszmajsak) [inline kubernetes/kubernetes/src/main/java/org/arquillian/cube/kubernetes/impl/ClientCreator.java]: Detail - if it's fluent  - let's chain.

File: openshift/openshift/src/main/java/org/arquillian/cube/openshift/impl/client/CubeOpenShiftConfiguration.java [hunk 8]
```diff
@@ -285,11 +269,12 @@ public int getRouterSniPort() {
...
 
-        if ((token == null || token.isEmpty()) && (client != null)) {
-            token = client.getClientExt().getConfiguration().getOauthToken();
+        if ((super.getToken() == null || super.getToken().isEmpty()) && (client != null)) {
+          String token = client.getClientExt().getConfiguration().getOauthToken();
+          setToken(token);
         }
 
-        return token;
+        return super.getToken();
```
Comment (bartoszmajsak) [inline openshift/openshift/src/main/java/org/arquillian/cube/openshift/impl/client/CubeOpenShiftConfiguration.java]: Why `super.` here? Do we have any extra logic above to be concerned about?

File: kubernetes/kubernetes/src/main/java/org/arquillian/cube/kubernetes/api/Configuration.java (hunk 6 above)
Comment (dipak-pawar) [reply kubernetes/kubernetes/src/main/java/org/arquillian/cube/kubernetes/api/Configuration.java]: We are looking for system property and env variable. If we used `username`, then `System.getenv("USERNAME")` returns your host name. 

I tried `user.name` as well, but `System.getProperty("user.name")` returns your host name. thats why I am using `name`

Comment (bartoszmajsak) [reply kubernetes/kubernetes/src/main/java/org/arquillian/cube/kubernetes/api/Configuration.java]: How about we prefix everything with `arq.cube` to have our own namespace? @lordofthejars wdyt?

File: docs/kubernetes.adoc (hunk 1 above)
Comment (dipak-pawar) [reply docs/kubernetes.adoc]: We are looking for system property and env variable. If we used username, then `System.getenv("USERNAME")` returns your host name.

I tried user.name as well, but `System.getProperty("user.name")` returns your host name. thats why I am using name

Comment (lordofthejars) [reply docs/kubernetes.adoc]: ok, then leave it as is.

File: kubernetes/kubernetes/src/main/java/org/arquillian/cube/kubernetes/api/Configuration.java (hunk 6 above)
Comment (lordofthejars) [reply kubernetes/kubernetes/src/main/java/org/arquillian/cube/kubernetes/api/Configuration.java]: I don't know. I am neither so sure about it, maybe leave it as Diapk is the best approach now.

File: openshift/openshift/src/main/java/org/arquillian/cube/openshift/impl/client/CubeOpenShiftConfiguration.java (hunk 8 above)
Comment (dipak-pawar) [reply openshift/openshift/src/main/java/org/arquillian/cube/openshift/impl/client/CubeOpenShiftConfiguration.java]: Using `super.` covers both scenario. i.e.
1. When there is no token, with client not null, it should get token from config and set it. and then return it.
2.  If token is available in parent, then return it.

File: openshift/openshift/src/main/java/org/arquillian/cube/openshift/impl/client/OpenShiftClientCreator.java (hunk 3 above)
Comment (dipak-pawar) [reply openshift/openshift/src/main/java/org/arquillian/cube/openshift/impl/client/OpenShiftClientCreator.java]: @lordofthejars I cannot see any parameters for authentication in your PR

File: kubernetes/kubernetes/src/main/java/org/arquillian/cube/kubernetes/api/Configuration.java (hunk 6 above)
Comment (dipak-pawar) [reply kubernetes/kubernetes/src/main/java/org/arquillian/cube/kubernetes/api/Configuration.java]: I will keep it the same for now.

**Comment by lordofthejars** [comment]
About parameter names, I think we need to use some time in next cabal to unify all parameters docker, openshift and kubernetes, and we create an issue, and when we want to release  2.0.0 of Cube, then let's refactor all of them.

**Comment by lordofthejars** [comment]
Good job @dipak-pawar Now we have fixed an old bug so now we are able to authenticate using token
